### PR TITLE
feat(postcss-prefix-selector): add types for v1.14

### PR DESCRIPTION
**Pull request by robertmaier**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

**Comment by robertmaier** [comment]
I have a question regarding the failing build.
The postcss-prefix-selector uses a postcss type internally. Postcss comes with it's own types, which are not part of DT.

Is there any way I can still import and use these types in my PR?
If not, I'm afraid I'd have to use an `any` instead

File: types/postcss-prefix-selector/tsconfig.json [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+        "strictFunctionTypes": true,
+        "strictNullChecks": true,
+        "baseUrl": "../",
+        "typeRoots": [
+            "../"
+        ],
+        "types": [],
+        "noEmit": true,
+        "forceConsistentCasingInFileNames": true,
+        "esModuleInterop": true
```
Comment (peterblazejewicz) [inline types/postcss-prefix-selector/tsconfig.json:18]:
```suggestion
        "forceConsistentCasingInFileNames": true
```
why this option?
https://github.com/DefinitelyTyped/DefinitelyTyped#esmoduleinteropallowsyntheticdefaultimports

File: types/postcss-prefix-selector/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
+// Type definitions for postcss-prefix-selector 1.14
```
Comment (peterblazejewicz) [inline types/postcss-prefix-selector/index.d.ts:1]:
```suggestion
// Type definitions for postcss-prefix-selector 1.15
```
https://github.com/RadValentin/postcss-prefix-selector/releases/tag/v1.15.0
(released)

File: types/postcss-prefix-selector/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+// Project: https://github.com/RadValentin/postcss-prefix-selector
+// Definitions by: robertmaier <https://github.com/robertmaier>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped/types/postcss-prefix-selector
+
+interface Options {
+    prefix?: string;
+    exclude?: ReadonlyArray<string | RegExp>;
+    ignoreFiles?: ReadonlyArray<string | RegExp>;
+    includeFiles?: ReadonlyArray<string | RegExp>;
+    transform?: (prefix: Readonly<string>, selector: Readonly<string>, prefixedSelector?: Readonly<string>, file?: Readonly<string>) => string;
```
Comment (peterblazejewicz) [inline types/postcss-prefix-selector/index.d.ts:11]:
```suggestion
    prefix?: string | undefined;
    exclude?: ReadonlyArray<string | RegExp> | undefined;
    ignoreFiles?: ReadonlyArray<string | RegExp> | undefined;
    includeFiles?: ReadonlyArray<string | RegExp> | undefined;
    transform?: (prefix: Readonly<string>, selector: Readonly<string>, prefixedSelector?: Readonly<string>, file?: Readonly<string>) => string | undefined;
```
do you want to support strict optional property (we are promoting this as kind of better interop with other libraries):
https://www.typescriptlang.org/play?exactOptionalPropertyTypes=true&ts=4.4.2&ssl=1&ssc=1&pln=2&pc=1#code/PTAEHUEsBcAtQKYA8CGBjaB5ADtSB7AOxQBsAFAJ32wQugE8AVemgZ0WICMSEATAGlDMaAZTQVIuALAAoEKFgp2raBIy1QFAK492KKlsK9QAd0XRQ9fFtBoUhUKwQX7oaniKlZ87FRp16UGh8ADpZbzAAMXwKRFQAW2weQThIdkhCaFoAM3QEUF4ENBJ9BHY4FAs4Wny00BQI0F9qWgZTWEg0eDsHTnyifuygk3xHVQyAc1YALlAAcl59AGs5t1i5kkgJ2Gg5xpjQGEdYaxJjQnwLPsOHardOACsi6DCZWQysily0fIBVJwoABEELkdNB2ABvWSgUDyRiwfIoThOQg-NxDFCgABupC0+QoCF8ZQQmXYc1Y9BUCHiexkMLQ+BIMXh1IQmCxtAkhQA-LMAESLChLPmgAA+oD5m220D5AG5ZABfcJyMBQODWKoddjZEoTDhIngCIIIgn1U1wAn5HEkPHsMydeBWGw9RpOCwAAwZTIoLPibI5FC5CHdQVGfX5guFgklWx2IvsxndhkK2QyfHdr0aImceEIeru1rxodASaMILTvBDJkgJBI9Vr+BMoHi+BUoAAUigcWIJLhNIY8H7Gl0ikttQc7sg0llUflgqBcjXBPauod2CDss9IBySIFciQKa95AAJRsIAMpE0Icn1BdaABe97TrHKo1S7D6iixBAoISECIKctCBgAhCEabIDk9RlmQRP12U5SBChDAB9Xh8DKQg5mgZCbmKLRChLZMgPTTMVX-fIdRQPV3WQdAsFwUDSEoFoAmEMoQ3iFAljKIQWAQHtJBcNA0C0ChKgQXdGk46BVzuFMMhAogmiob9CmMJR6gcdxGLrZp-AYUjCmKUo71RDwHAmZx-lobNoFzKYAAoAEpZmsoEQRQMFWHlN4ZAZQg2zdez2AAXlASzoDc2zgucnygsmVgQi9GDWXgwNEPyMKBWWOVZHivNEuSn1YP9BCCKyqU4x8xoACEtAsC5oGmPKcwSpLoOK1KAyDUAwqI1NCD4aqyMoH9izuSjqNojAcHMpi-FaJg+NYd1LzqEwNMappW1YSBuAQEIgA

Comment (peterblazejewicz) [inline types/postcss-prefix-selector/index.d.ts:11]: this one seems to be always called with all 4 params:
https://github.com/RadValentin/postcss-prefix-selector/blob/master/index.js#L43-L48
params for callbacks are always optional:
https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html#optional-parameters-in-callbacks

File: types/postcss-prefix-selector/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
+// Type definitions for postcss-prefix-selector 1.14
+// Project: https://github.com/RadValentin/postcss-prefix-selector
+// Definitions by: robertmaier <https://github.com/robertmaier>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped/types/postcss-prefix-selector
+
+interface Options {
+    prefix?: string;
```
Comment (peterblazejewicz) [inline types/postcss-prefix-selector/index.d.ts:7]: isn't the prefix required?
https://github.com/RadValentin/postcss-prefix-selector/blob/master/index.js#L2
(and by the very purpose `Prefix every CSS selector with a custom namespace .a => .prefix .a`)

File: types/postcss-prefix-selector/tsconfig.json [hunk 5]
```diff
@@ -15,7 +15,7 @@
         "types": [],
         "noEmit": true,
         "forceConsistentCasingInFileNames": true,
-        "esModuleInterop": true
+        "exactOptionalPropertyTypes": true
```
Comment (peterblazejewicz) [inline types/postcss-prefix-selector/tsconfig.json:18]: please leave config with defaults, we code for 'exactOptionalPropertyTypes' support, but without enforcing it via hardwired property.
`esModuleInterop` options should not be here either.

File: types/postcss-prefix-selector/postcss-prefix-selector-tests.ts [hunk 6]
```diff
@@ -1,4 +1,4 @@
 import postcssPrefixSelector from 'postcss-prefix-selector';
```
Comment (peterblazejewicz) [inline types/postcss-prefix-selector/postcss-prefix-selector-tests.ts:1]:
```suggestion
import postcssPrefixSelector ='require(postcss-prefix-selector');
```
IMO this is old school CJS module, with direct function symbol export:
https://github.com/RadValentin/postcss-prefix-selector/blob/master/index.js#L1

File: types/postcss-prefix-selector/index.d.ts [hunk 7]
```diff
@@ -1,16 +1,16 @@
...
+    exclude?: ReadonlyArray<string | RegExp> | undefined;
+    ignoreFiles?: ReadonlyArray<string | RegExp> | undefined;
+    includeFiles?: ReadonlyArray<string | RegExp> | undefined;
+    transform?: ((prefix: Readonly<string>, selector: Readonly<string>, prefixedSelector: Readonly<string>, file: Readonly<string>) => string) | undefined;
 }
 
 declare function postcssPrefixSelector(options: Readonly<Options>): (root: any) => string | undefined;
 
-export = postcssPrefixSelector;
+export default postcssPrefixSelector;
```
Comment (peterblazejewicz) [inline types/postcss-prefix-selector/index.d.ts:16]: I think that export was OK:
https://github.com/RadValentin/postcss-prefix-selector/blob/master/index.js#L1

File: types/postcss-prefix-selector/index.d.ts [hunk 8]
```diff
@@ -1,16 +1,16 @@
...
 // Definitions by: robertmaier <https://github.com/robertmaier>
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped/types/postcss-prefix-selector
 
 interface Options {
-    prefix?: string;
-    exclude?: ReadonlyArray<string | RegExp>;
-    ignoreFiles?: ReadonlyArray<string | RegExp>;
-    includeFiles?: ReadonlyArray<string | RegExp>;
-    transform?: (prefix: Readonly<string>, selector: Readonly<string>, prefixedSelector?: Readonly<string>, file?: Readonly<string>) => string;
+    prefix: string;
```
Comment (peterblazejewicz) [inline types/postcss-prefix-selector/index.d.ts:7]:
```suggestion
    prefix?: string | undefined;
```
I'm really sorry, I was wrong, this is not required option, I've misread. Sorry for invenvienance.

File: types/postcss-prefix-selector/postcss-prefix-selector-tests.ts (hunk 6 above)
Comment (robertmaier) [reply types/postcss-prefix-selector/postcss-prefix-selector-tests.ts]: that was the syntax i was looking for :) thanks for the suggestion, this did the trick

**Review by peterblazejewicz — APPROVED** [review]
LGTM!
@robertmaier thx!

**Comment by robertmaier** [comment]
Ready to merge

**Comment by robertmaier** [comment]
Thanks for the review @peterblazejewicz 👋

**Pull request merged**